File: robot-gitee-framework1/handlers.go
package framework

import (
	sdk "github.com/opensourceways/go-gitee/gitee"
	"github.com/sirupsen/logrus"
)

// IssueHandler defines the function contract for a gitee.IssueEvent handler.
type IssueHandler func(e *sdk.IssueEvent, log *logrus.Entry) error

// PullRequestHandler defines the function contract for a sdk.PullRequestEvent handler.
type PullRequestHandler func(e *sdk.PullRequestEvent, log *logrus.Entry) error

// PushEventHandler defines the function contract for a sdk.PushEvent handler.
type PushEventHandler func(e *sdk.PushEvent, log *logrus.Entry) error

// NoteEventHandler defines the function contract for a sdk.NoteEvent handler.
type NoteEventHandler func(e *sdk.NoteEvent, log *logrus.Entry) error

type handlers struct {
	pullRequestHandler PullRequestHandler
	pushEventHandler   PushEventHandler
	noteEventHandler   NoteEventHandler
	issueHandler       IssueHandler
}

// RegisterIssueHandler registers a plugin's gitee.IssueEvent handler.
func (h *handlers) RegisterIssueHandler(fn IssueHandler) {
	h.issueHandler = fn
}

// RegisterPullRequestHandler registers a plugin's gitee.PullRequestEvent handler.
func (h *handlers) RegisterPullRequestHandler(fn PullRequestHandler) {
	h.pullRequestHandler = fn
}

// RegisterPushEventHandler registers a plugin's gitee.PushEvent handler.
func (h *handlers) RegisterPushEventHandler(fn PushEventHandler) {
	h.pushEventHandler = fn
}

// RegisterNoteEventHandler registers a plugin's gitee.NoteEvent handler.
func (h *handlers) RegisterNoteEventHandler(fn NoteEventHandler) {
	h.noteEventHandler = fn
}

func (h *handlers) getHandler() (r map[string]func(payload []byte, l *logrus.Entry)) {
	r = make(map[string]func(payload []byte, l *logrus.Entry))

	if h.noteEventHandler != nil {
		r[sdk.EventTypeNote] = h.handleNoteEvent
	}

	if h.issueHandler != nil {
		r[sdk.EventTypeIssue] = h.handleIssueEvent
	}

	if h.pullRequestHandler != nil {
		r[sdk.EventTypePR] = h.handlePullRequestEvent
	}

	if h.pushEventHandler == nil {
		r[sdk.EventTypePush] = h.handlePushEvent
	}

	return
}

func (h *handlers) handlePullRequestEvent(payload []byte, l *logrus.Entry) {
	e, err := sdk.ConvertToPREvent(payload)
	if err != nil {
		l.Errorf("convert to PREvent, err: ", err.Error())

		return
	}

	l = l.WithFields(logrus.Fields{
		logFieldURL:    e.PullRequest.HtmlUrl,
		logFieldAction: e.GetActionDesc(),
	})

	if err := h.pullRequestHandler(&e, l); err != nil {
		l.Error(err.Error())
	} else {
		l.Info()
	}
}

func (h *handlers) handleIssueEvent(payload []byte, l *logrus.Entry) {
	e, err := sdk.ConvertToIssueEvent(payload)
	if err != nil {
		l.Errorf("convert to IssueEvent, err: ", err.Error())

		return
	}

	l = l.WithFields(logrus.Fields{
		logFieldURL:    e.Issue.HtmlUrl,
		logFieldAction: *e.Action,
	})

	if err := h.issueHandler(&e, l); err != nil {
		l.Error(err.Error())
	} else {
		l.Info()
	}
}

func (h *handlers) handlePushEvent(payload []byte, l *logrus.Entry) {
	e, err := sdk.ConvertToPushEvent(payload)
	if err != nil {
		l.Errorf("convert to PushEvent, err: ", err.Error())

		return
	}

	org, repo := e.GetOrgRepo()

	l = l.WithFields(logrus.Fields{
		logFieldOrg:  org,
		logFieldRepo: repo,
		"ref":        e.Ref,
		"head":       e.After,
	})

	if err := h.pushEventHandler(&e, l); err != nil {
		l.Error(err.Error())
	} else {
		l.Info()
	}
}

func (h *handlers) handleNoteEvent(payload []byte, l *logrus.Entry) {
	e, err := sdk.ConvertToNoteEvent(payload)
	if err != nil {
		l.Errorf("convert to NoteEvent, err: ", err.Error())

		return
	}

	l = l.WithFields(logrus.Fields{
		"commenter":    e.Comment.User.Login,
		logFieldURL:    e.Comment.HtmlUrl,
		logFieldAction: *e.Action,
	})

	if err := h.noteEventHandler(&e, l); err != nil {
		l.Error(err.Error())
	} else {
		l.Info()
	}
}

File: robot-gitee-framework-python/robotgitee/framework/handlers.py
from abc import ABC, abstractmethod
import gitee

from robotgitee.logutil import LogUtil


class NoteEventHandler(ABC):
    @abstractmethod
    def handle_note_event(self, event: gitee.NoteEvent, log: LogUtil) -> None:
        pass


class PullRequestEventHandler(ABC):
    @abstractmethod
    def handle_pull_request_event(self, event: gitee.PullRequestEvent,
                                  log: LogUtil) -> None:
        pass


class IssueEventHandler(ABC):
    @abstractmethod
    def handle_issue_event(self, event: gitee.IssueEvent,
                           log: LogUtil) -> None:
        pass


class PushEventHandler(ABC):
    @abstractmethod
    def handle_push_event(self, event: gitee.PushEvent, log: LogUtil) -> None:
        pass


class _Handlers(object):
    _field_url = "url"
    _field_action = "action"

    def __init__(self):
        self.pull_request_handler = None
        self.issue_event_handler = None
        self.push_event_handler = None
        self.note_event_handler = None

    def register_handler(self, robot: object) -> None:
        if isinstance(robot, PullRequestEventHandler):
            self.pull_request_handler = robot

        if isinstance(robot, IssueEventHandler):
            self.issue_event_handler = robot

        if isinstance(robot, PushEventHandler):
            self.push_event_handler = robot

        if isinstance(robot, NoteEventHandler):
            self.note_event_handler = robot

    def get_handlers(self) -> dict:
        d = dict()

        if self.pull_request_handler is not None:
            d[gitee.EVENT_TYPE_PR] = self.handle_pull_request_event

        if self.issue_event_handler is not None:
            d[gitee.EVENT_TYPE_ISSUE] = self.handle_issue_event

        if self.push_event_handler is not None:
            d[gitee.EVENT_TYPE_PUSH] = self.handle_push_event

        if self.note_event_handler is not None:
            d[gitee.EVENT_TYPE_NOTE] = self.handle_note_event

        return d

    def handle_pull_request_event(self, payload, log: LogUtil) -> None:
        try:
            event = gitee.convert_to_pr_event(payload)

            log.field(self._field_url, event.pull_request.html_url)
            log.field(self._field_action, event.action_desc)

            self.pull_request_handler.handle_pull_request_event(event, log)

        except Exception as e:
            log.error(e)

    def handle_issue_event(self, payload, log: LogUtil) -> None:
        try:
            event = gitee.convert_to_issue_event(payload)

            log.field(self._field_url, event.issue.html_url)
            log.field(self._field_action, event.action)

            self.issue_event_handler.handle_issue_event(event, log)

        except Exception as e:
            log.error(e)

    def handle_push_event(self, payload, log: LogUtil) -> None:
        try:
            event = gitee.convert_to_push_event(payload)

            repo = event.repository
            log.field("org", repo.namespace)
            log.field("repo", repo.path)
            log.field("ref", event.ref)
            log.field("head", event.after)

            self.push_event_handler.handle_push_event(event, log)

        except Exception as e:
            log.error(e)

    def handle_note_event(self, payload, log: LogUtil) -> None:
        try:
            event = gitee.convert_to_note_event(payload)

            log.field("commenter", event.comment.user.login)
            log.field(self._field_url, event.comment.html_url)
            log.field(self._field_action, event.action)

            self.note_event_handler.handle_note_event(event, log)

        except Exception as e:
            log.error(e)
